File: backend/app/schemas/report.py
import uuid
from datetime import date
from typing import Literal

from pydantic import BaseModel, Field

from app.models.transaction import TransactionCategory


class CategoryAmount(BaseModel):
    category: TransactionCategory
    category_label: str
    amount: float
    share_pct: float


class MonthlySummary(BaseModel):
    year: int
    month: int
    total_income: float
    total_expense: float
    net: float
    transaction_count: int
    income_by_category: list[CategoryAmount] = Field(default_factory=list)
    expense_by_category: list[CategoryAmount] = Field(default_factory=list)


class InsightItem(BaseModel):
    level: Literal["critical", "warning", "good", "info"]
    title: str
    message: str
    action: str


class AnomalyItem(BaseModel):
    transaction_id: uuid.UUID | None = None
    transaction_date: date
    category: TransactionCategory
    category_label: str
    amount: float
    description: str | None = None
    anomaly_score: float
    method: str
    reason: str


class TrendData(BaseModel):
    slope: float
    direction: Literal["up", "down", "stable", "insufficient_data"]


class VolatilityData(BaseModel):
    cv: float | None
    mean: float


class ForecastData(BaseModel):
    forecast_total: float | None
    daily_forecast: list[float] = Field(default_factory=list)
    residual_std: float | None = None
    confidence: str


class CashRunway(BaseModel):
    runway_days: float | None
    avg_daily_net: float



class ReportResponse(BaseModel):
    summary: MonthlySummary
    income_volatility: VolatilityData | None = None
    trend: TrendData | None = None
    anomalies: list[AnomalyItem] = Field(default_factory=list)
    forecast: ForecastData | None = None
    cash_runway: CashRunway | None = None
    insights: list[InsightItem] = Field(default_factory=list)

File: frontend/lib/types/report.ts
export type InsightLevel = "critical" | "warning" | "good" | "info";

export type TrendDirection =
  | "tang"
  | "giam"
  | "on_dinh"
  | "khong_du_du_lieu";

export interface CategoryAmount {
  category: string;
  category_label: string;
  amount: number;
  share_pct: number;
}

export interface MonthlySummary {
  year: number;
  month: number;
  total_income: number;
  total_expense: number;
  net: number;
  transaction_count: number;

  income_by_category: CategoryAmount[];
  expense_by_category: CategoryAmount[];
}

export interface InsightItem {
  level: InsightLevel;
  title: string;
  message: string;
  action: string;
}

export interface AnomalyItem {
  transaction_id: string | null;
  transaction_date: string;

  category: string;
  category_label: string;

  amount: number;
  description: string | null;

  anomaly_score: number;
  method: string;
  reason: string;
}

export interface TrendData {
  slope: number;
  direction: TrendDirection;
}

export interface VolatilityData {
  cv: number | null;
  mean: number;
}

export interface ForecastData {
  forecast_total: number | null;
  daily_forecast: number[];

  residual_std: number | null;
  confidence: string;
}

export interface CashRunway {
  runway_days: number | null;
  avg_daily_net: number;
}

export interface ReportResponse {
  summary: MonthlySummary;

  income_volatility: VolatilityData | null;

  trend: TrendData | null;

  anomalies: AnomalyItem[];

  forecast: ForecastData | null;

  cash_runway: CashRunway | null;

  insights: InsightItem[];
}